<a href="https://colab.research.google.com/github/ddutta111/Debolina_Projects/blob/main/Churn_Modelling_Bank_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Business Problem Description**

An European bank has noticed a concerning increase in customer churn, with more customers leaving than expected. To get to the root of this issue, they conducted a study six months ago by selecting a random sample of 10,000 customers from their millions of clients across Europe. They collected detailed data on their demographic and financial characteristics, such as customer ID, credit score, geography, gender, age, tenure, account balance, product holdings, credit card ownership, active membership status, and estimated salary.After monitoring these customers for six months, the bank identified who among them stayed and who left, indicated by a binary outcome in the dataset. They are now looking for expertise of a data analyst to help them analyze this dataset.

# **Business Goal**

As an analyst, my primary goal is to develop a geodemographic segmentation model to predict which customers are most at risk of leaving the bank. This analysis will empower the bank to implement targeted strategies aimed at retaining these customers.

# **Business Variable & Modelling Explanation**

Model Overview:

Here, I have implemented a classic Artificial Neural Network (ANN), specifically a fully-connected neural network. This model consists of only fully-connected layers, with no convolutional or specialized layers. The goal is to predict a binary outcome, which in this case is whether a customer will leave the bank (churn) or stay. ANN models are versatile and can be used for both regression and classification tasks. In this project, I will focus on ANN with classification.

**Dataset and Variables:**

The dataset consists of information collected by the bank about its customers. Below are the details of the variables:

RowNumber: Non-relevant, sequential identifier of rows in the dataset (will be excluded).

CustomerID: Unique identification key for each customer.

Surname: The last name of the customer (irrelevant for prediction).

CreditScore: Numerical score representing the creditworthiness of the customer.

Geography: The country where the customer resides (e.g., France, Germany, Spain).

Gender: The gender of the customer (Male/Female).

Age: Age of the customer.

Tenure: Number of years the customer has been with the bank.

Balance: The current account balance of the customer.

NumOfProducts: The number of bank products the customer is using (e.g., credit card, loan).

HasCrCard: Binary indicator of whether the customer has a credit card (1 = Yes, 0 = No).

IsActiveMember: Binary indicator of whether the customer is considered active (1 = Active, 0 = Inactive).

EstimatedSalary: Estimated annual salary of the customer.

Exited: Dependent variable indicating whether the customer has left the bank (1 = Yes, 0 = No).

# Artificial Neural Network

**ANN Model Objective:**

The objective of this project is to develop an Artificial Neural Network (ANN) model that predicts the likelihood of customer churn (whether a customer will leave the bank). By identifying patterns and correlations in the dataset, the model will enable the bank to take proactive measures to retain at-risk customers.

**Methodology Overview:**

*Part One: Data Pre-processing*

Step 1: Import the dataset.

Step 2: Perform data pre-processing, including standardization, encoding categorical variables, and splitting the data into training and test sets.

*Part Two: Building the ANN*

Step 1: Initialize the ANN.

Step 2: Add the input layer and the first hidden layer.

Step 3: Add the second hidden layer.

Step 4: Add the output layer.

*Part Three: Training the ANN*

Step 1: Compile the ANN with an appropriate optimizer and loss function.

Step 2: Train the ANN on the training dataset.

*Part Four: Making Predictions and Model Evaluation*

Step 1: Deploy the model to predict whether a new customer will stay or leave the bank.

Step 2: Predict the outcomes on the test set to generate predictions.

Step 3: Evaluate model performance by constructing a confusion matrix and calculating accuracy.

### Importing the libraries

In [1]:
import numpy as np
import pandas as pd
import tensorflow as tf

In [2]:
tf.__version__

'2.17.0'

## Part 1 - Data Preprocessing

### Importing the dataset

In [3]:
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:-1].values
y = dataset.iloc[:, -1].values

In [4]:
print(X)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


In [5]:
print(y)

[1 0 1 ... 1 1 0]


### Encoding categorical data

Label Encoding the "Gender" column

In [6]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X[:, 2] = le.fit_transform(X[:, 2])

In [7]:
print(X)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


One Hot Encoding the "Geography" column

In [8]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder',OneHotEncoder(), [1])], remainder= 'passthrough')
X = np.array(ct.fit_transform(X))

In [9]:
print(X)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


### Splitting the dataset into the Training set and Test set

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size = 0.2, random_state = 1)

### Feature Scaling

In [11]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Part 2 - Building the ANN

### Initializing the ANN

In [12]:
ann = tf.keras.models.Sequential()

### Adding the input layer and the first hidden layer

In [13]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the second hidden layer

In [14]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the output layer

In [15]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

## Part 3 - Training the ANN

### Compiling the ANN

In [16]:
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

### Training the ANN on the Training set

In [17]:
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)

Epoch 1/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.4494 - loss: 0.7810
Epoch 2/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7966 - loss: 0.5078
Epoch 3/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8156 - loss: 0.4460
Epoch 4/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8259 - loss: 0.4246
Epoch 5/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.8163 - loss: 0.4313
Epoch 6/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8305 - loss: 0.4063
Epoch 7/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.8204 - loss: 0.4158
Epoch 8/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.8340 - loss: 0.3891
Epoch 9/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8360 - loss: 0.3857
Epoch 10/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.8419 - loss: 0.3826
Epoch 11/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.8446 - loss: 0.3635
Epoch 12/100
250/250 ━━━━━━━━━━━━━━━━━━━━

# **Result**

Here, the ANN model converged to an accuracy of 0.86 by the 20th epoch, indicating strong performance. Training completed in about 20 seconds, with the model correctly predicting 86 out of 100 observations on the training set. We will need to verify this accuracy on the test set for further validation.

## Part 4 - Making the predictions and evaluating the model

### Predicting the result of a single observation

The model is capable of predicting the outcome for a single customer, determining whether they are likely to stay with or leave the bank. This prediction is based on the customer's specific data, using the trained ANN model to assess their churn risk.

**Homework**

Use our ANN model to predict if the customer with the following informations will leave the bank:

Geography: France

Credit Score: 600

Gender: Male

Age: 40 years old

Tenure: 3 years

Balance: \$ 60000

Number of Products: 2

Does this customer have a credit card? Yes

Is this customer an Active Member: Yes

Estimated Salary: \$ 50000

So, should we say goodbye to that customer?

**Solution**

In [23]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])))

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 99ms/step
[[0.02974912]]


In [24]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]]))> 0.5)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step
[[False]]


# **Result Explanation:**

Here, we see the predicted probability of customer leaing the bank is 0.03 which is < 0.5. Since, If predicted probability of customer leaving the bank is > 0.5, then, its "True", Otherwise "False".

**Therefore, our ANN model predicts that this customer stays in the bank!**

**Important note 1:** Notice that the values of the features were all input in a double pair of square brackets. That's because the "predict" method always expects a 2D array as the format of its inputs. And putting our values into a double pair of square brackets makes the input exactly a 2D array.

**Important note 2:** Notice also that the "France" country was not input as a string in the last column but as "1, 0, 0" in the first three columns. That's because the predict method expects the one-hot-encoded values of the state, and as we see in the first row of the matrix of features X, "France" was encoded as "1, 0, 0". And be careful to include these values in the first three columns, because the dummy variables are always created in the first columns.

### Predicting the Test set results

In [25]:
y_pred = ann.predict(X_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step
[[0 0]
 [0 0]
 [0 0]
 ...
 [0 0]
 [0 0]
 [1 0]]


### Making the Confusion Matrix

In [26]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[1532   53]
 [ 216  199]]


0.8655

# **Conclusion**

The model performed well on the test set, demonstrating strong predictive capability. For example, the first three customers in the test set were correctly predicted to stay in the bank, matching their actual outcomes. Similarly, in the final set, the first and second customers were accurately predicted to stay, while the model incorrectly predicted that the last customer would leave, though they actually stayed.

Overall, the results are promising, with the confusion matrix showing an accuracy of over 86.7%. This indicates that out of 100 customers, 87 were correctly predicted to either stay with or leave the bank. Specifically, the model made 1,532 correct predictions that customers would stay, 199 correct predictions that they would leave, 53 incorrect predictions that customers would leave when they actually stayed, and 216 incorrect predictions that customers would stay when they actually left. These results suggest the model is reliable and effective in predicting customer churn, offering valuable insights for the bank to retain customers.

# **Recommendations**

Based on the results of Churn_Modelling (by applying Ann Classsification) as an analyst following recommendations can be given to the bank:

1. **Targeted Retention Strategies:** Since the model accurately identifies customers at risk of leaving, the bank should implement targeted retention strategies for those predicted to churn. This could include personalized offers, improved customer service, or loyalty programs to encourage these customers to stay.

2. **Monitor and Adjust:** Regularly monitor the model’s predictions and compare them with actual outcomes. This will allow the bank to adjust its strategies and improve the model over time, ensuring it continues to accurately predict churn.

3. **Focus on High-Risk Segments:** Analyze the characteristics of the customers incorrectly predicted to stay but who actually left (false negatives). Understanding these cases can help the bank refine its approach to those segments, potentially reducing future churn.

4. **Leverage Success:** Utilize the high accuracy of the model to confidently allocate resources towards customers who are most likely to stay, optimizing marketing and customer service efforts where they will have the greatest impact.

5. **Continual Model Refinement:** As customer behaviors and market conditions evolve, it’s essential to regularly retrain and update the model with new data to maintain its effectiveness.

By following these recommendations, the European bank can reduce customer churn, enhance customer satisfaction, and improve overall business performance.






